xcdat/climatology.py

```diff
@@ -0,0 +1,217 @@
+    variables are also preserved since individual data variables are grouped,
+    rather than the entire dataset.
+
+    Once grouping is complete, attributes related to the operation are added to
+    the dataset for reference.
+
+    :param ds: A Dataset object
+    :type ds: xr.Dataset
+    :param period: The period of time to group data by
+    :type period: Period
+    :param is_weighted: Calculate weighted averages, defaults to True
+    :type is_weighted: bool, optional
+    :return: The calculated climatology
+    :rtype: xr.Dataset
+    """
+    time_period: TimePeriod = f"time.{period}"  # type:ignore
+    weights = _calculate_weights(ds, time_period) if is_weighted else None
+
+    for key in ds.data_vars.keys():
+        if "_bnds" not in str(key):
+            data_var = ds[key]
+
+            if is_weighted:
+                data_var *= weights
+
+            ds[key] = data_var.groupby(time_period).sum(dim="time")
+
+    ds.attrs.update(
+        {
+            "calculation_info": {
+                "type": calculation_type,
+                "period": period,
+                "is_weighted": is_weighted,
+            },
+        }
+    )
+    return ds
+
+
+def _calculate_weights(ds: xr.Dataset, time_period: TimePeriod) -> xr.DataArray:
+    """Calculates weights for a Dataset based on a period of time.
+
+    Leap years and number of days for each month are considered during grouping.
+
+    :param ds: A Dataset object
+    :type ds: xr.Dataset
+    :param time_period: The period of time for calculating weightes
+    :type time_period: TimePeriod
+    :return: The calculated weights based on the period of time
+    :rtype: xr.DataArray
+    """
+    months_lengths = _get_months_lengths(ds)
+
+    weights: xr.DataArray = (
+        months_lengths.groupby(time_period) / months_lengths.groupby(time_period).sum()
+    )
+    _validate_weights(ds, weights, time_period)
+
+    return weights
+
+
+def _get_months_lengths(ds: xr.Dataset) -> xr.DataArray:
+    """Get the lengths of the months based on the time coordinates.
+
+    If time bounds exist, they will be used to generate lengths of months. This
+    is the robust way because different datasets could record their time
+    differently (e.g., at beginning/end/middle of each time interval).
+
+    Otherwise, use the time variable, which may be less accurate.
+
+    :param ds: A Dataset object
+    :type ds: xr.Dataset
+    :return: The lengths of months
+    :rtype: xr.DataArray
+    """
+    time_bounds = ds.get("time_bnds")
+
+    if time_bounds is not None:
+        logging.info("Using existing time bounds to calculate weights.")
+        months_lengths = (time_bounds[:, 1] - time_bounds[:, 0]).dt.days
+    else:
+        # TODO: Generate time bounds if they don't exist?
+        logging.info("No time bounds found, using time to calculate weights.")
+        months_lengths = ds.time.dt.days_in_month
+
+    return months_lengths
```

Comment:
This function gets the lengths of the months for calculating weights by using either the time bounds or the time variable.

@chengzhuzhang  When time bounds don't exist, should the fallback behavior be to generate them, or stick to using the time variable? There isn't a handle for this in https://github.com/E3SM-Project/e3sm_diags/blob/master/acme_diags/driver/utils/climo.py#L37-L39.